text/0055-remove-unspecified.md

```diff
@@ -0,0 +1,248 @@
+
+```rust
+{
+    principal: Some(User::"alice"),
+    action: Some(Action::"createFile"),
+    resource: None,
+    ...
+}
+```
+
+In the schema, the only way to say that an action applies to an unspecified entity is by omitting the relevant `appliesTo` field.
+For example, the following schemas specify that the "createFile" action applies (only) to an unspecified resource.
+
+Cedar schema syntax ([RFC 24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md)):
+
+```
+action createFile
+  appliesTo { principal: [User] };
+```
+
+Cedar schema JSON syntax:
+
+```json
+"createFile": {
+    "appliesTo": {
+        "principal": ["User"]
+    }
+}
+```
+
+We have found this _implicit_ way of referring to unspecified entities (from this point on called "default entities") to be confusing, so this RFC proposes to introduce a new special type (`__cedar::Default`) that allows users to refer to default entities _explicitly_.
+
+This requires changing the type of the `Request` constructor from:
+
+```rust
+pub fn new(
+        principal: Option<EntityUid>,
+        action: Option<EntityUid>,
+        resource: Option<EntityUid>,
+        context: Context,
+        schema: Option<&Schema>,
+    ) -> Result<Self, RequestValidationError>
+...
+```
+
+to:
+
+```rust
+pub enum RequestEntity {
+    Default,
+    Unknown,
+    EntityUid(EntityUid)
+}
+
+pub fn new(
+        principal: RequestEntity,
+        action: RequestEntity,
+        resource: RequestEntity,
+        context: Context,
+        schema: Option<&Schema>,
+    ) -> Result<Self, RequestValidationError>
+...
+```
+
+(Note: the `Unknown` variant may be hidden behind the "partial-eval" experimental flag, depending on when this RFC is implemented.)
+
+Furthermore, schemas will need to explicitly say that an action applies to the default type, rather than omitting an `appliesTo` field.
+For example, the schemas above would need to be modified as follows.
+
+Cedar schema syntax (RFC 24):
+
+```
+action createFile
+  appliesTo { principal: [User], resource: [__cedar::Default] };
+```
+
+Cedar schema JSON syntax:
+
+```json
+"createFile": {
+    "appliesTo": {
+        "principal": ["User"],
+        "resource": ["__cedar::Default"]
+    }
+}
+```
+
+Users will not be allowed to include entities of type `__cedar::Default` in the store, which ensures that these entities will never have attributes or ancestors/descendants, which are properties we will rely on during validation.
+Users will not be allowed to reference the type `__cedar::Default` in their policies (so neither `principal == __cedar::Default::"principal"` nor `principal is __cedar::Default` are allowed).
+Users do not need to add `__cedar::Default` to their schemas (and it will be an error if they do) because `__cedar::Default` is defined automatically.
+Finally, users cannot create entity uids with type `__cedar::Default`, which prevents using default entities with the `RequestEntity::EntityUid` constructor.
```

Comment:
What is the goal of these restrictions? Is there a property we can state and prove about them?

I think that the goal is to prevent users from hijacking that implementation mechanism of default entities which would now be explicitly exposed. When the mechanism is fully hidden, as today, we don't have to worry about a user trying to make a new kind of default entity, or giving it ancestors, or referring to it in policy expressions. But now that we've given it an explicit name, there's mischief involving that name we have to prevent.

I originally thought the goal of the restrictions was different: To ensure that policies that might use the default entity _ignore_ it, since the point is that the entity _doesn't matter_. But these restrictions don't do that. You can still write a policy with a clause like `principal == User::"Alice"` and if you pass in the default entity for the principal then this expression will _not_ be true, affecting policy satisfaction. You could also write `principal has manager && principal.manager == User::"Bob"` and again this will evaluate just fine with the default entity. This feels weird. Further up we write

> If a policy uses an _unknown entity_ in a non-trivial way, then evaluation will produce a residual.
> If a policy uses an _unspecified entity_ in a non-trivial way, then evaluation will produce an error.

The caveat "non-trivial" seems strange. What makes behavior trivial? If the use of a default entity affects the policy evaluation outcome in _any_ way, that seems "non trivial".

It would make more sense to me to restrict _all_ uses of default entities. For example, if `Action::"createFile"` was specified to as having a default resource entity (the schema would say `createFile` applies to `__cedar::Default`), then a policy that could use `Action::"createFile"` as its action should always ignore the `resource` component. If it doesn't, that's an error. In other words, if you wrote the following, you should get an evaluation and validation error for the expression `resource == File::"Root"`:
```
permit(principal, action == Action::"createFile", resource == File::"Root")
```
This is a more invasive change, so probably we don't want to do it. But at the same time I think that sheds some light on the confusion around unspecified entities today. This RFC is not proposing to change the semantics of unspecified entities make them less confusing, it is just changing the implicit specification we have now to be a (equally confusing) explicit one.